connectors/byoei.py

```diff
@@ -409,17 +444,61 @@ async def get_existing_ids(self, index):
             ts = doc["_source"].get(TIMESTAMP_FIELD)
             yield doc_id, ts
 
+    def done(self):
+        if self._fetcher_task is not None and not self._fetcher_task.done():
+            return False
+        if self._bulker_task is not None and not self._bulker_task.done():
+            return False
+        return True
+
+    async def cancel(self):
+        if self._fetcher_task is not None:
+            self._fetcher_task.cancel()
+            await self._fetcher_task
+        if self._bulker_task is not None:
+            self._bulker_task.cancel()
+            await self._bulker_task
+
+    def ingestion_stats(self):
+        stats = {}
+        if self._fetcher is not None:
+            stats.update(
+                {
+                    "doc_created": self._fetcher.total_docs_created,
+                    "attachment_extracted": self._fetcher.total_downloads,
+                    "doc_updated": self._fetcher.total_docs_updated,
+                    "doc_deleted": self._fetcher.total_docs_deleted,
+                }
+            )
+        if self._bulker is not None:
+            stats.update(
+                {
+                    "bulk_operations": dict(self._bulker.ops),
+                    "indexed_document_count": self._bulker.indexed_document_count,
+                    "indexed_document_volume": round(
+                        self._bulker.indexed_document_volume
+                    ),
+                    "deleted_document_count": self._bulker.deleted_document_count,
+                }
+            )
+        return stats
+
+    def fetch_error(self):
+        return None if self._fetcher is None else self._fetcher.fetch_error
+
     async def async_bulk(
         self,
         index,
         generator,
         pipeline,
-        filter=None,
+        filter_=None,
```

Comment:
do this in another PR

Reply:
Done here: https://github.com/elastic/connectors-python/pull/525